src/python/pants/rules/core/test.py

```diff
@@ -25,8 +24,11 @@ class Test(Goal):
 
 
 @console_rule
-def fast_test(console: Console, addresses: BuildFileAddresses) -> Test:
-  test_results = yield [Get(TestResult, Address, address.to_address()) for address in addresses]
+def fast_test(console: Console, addresses: BuildFileAddresses,
+              build_config: BuildConfiguration) -> Test:
+  all_targets = yield Get(HydratedTargets, BuildFileAddresses, addresses)
+  filtered_targets = build_config.union_members(TestTarget, all_targets)
```

Comment:
I mildly prefer:
```python
test_results = yield [
  Get(TestResult, HydratedTarget, tgt)
  for tgt in targets
  if build_config.is_union_member(TestTarget, tgt) # or tgt.adaptor
]
```

but not enough to argue about it if you don't :)

Reply:
Can do, but each call to `is_union_member` would end up being a linear scan of the (probably very short) list of union members for the union base. Unless we change the data structures inside BuildConfiguration.

Reply:
Oh, hm, looks like it's already an OrderedSet in practice.

Reply:
And just realized we need the list of filtered targets to zip with the test_results (my current code in this PR is broken there).